### PR TITLE
STITCH-914 Add custom auth to JavaScript SDK

### PR DESCRIPTION
**Pull request by jsflax**

### CODE REVIEW

File: src/auth/providers.js [hunk 1]
```diff
@@ -54,6 +54,36 @@ function anonProvider(auth) {
...
 }
 
+function customProvider(auth) {
+  const providerRoute = 'providers/custom-token';
+  const loginRoute = `${providerRoute}/login`;
+  return {
+    /**
+     * Login to a stitch application using username and password authentication
+     *
+     * @memberof userPassProvider
```
Comment (edaniels) [inline src/auth/providers.js]: This docstring is incorrect

File: src/auth/providers.js [hunk 2]
```diff
@@ -54,6 +54,36 @@ function anonProvider(auth) {
   };
 }
 
+function customProvider(auth) {
+  const providerRoute = 'providers/custom-token';
+  const loginRoute = `${providerRoute}/login`;
+  return {
```
Comment (edaniels) [inline src/auth/providers.js]: newline before this line

File: src/auth/providers.js [hunk 3]
```diff
@@ -331,7 +361,8 @@ function createProviders(auth, options = {}) {
     google: googleProvider(auth),
     facebook: facebookProvider(auth),
     mongodbCloud: mongodbCloudProvider(auth),
-    userpass: userPassProvider(auth)
+    userpass: userPassProvider(auth),
+    custom: customProvider(auth)
```
Comment (edaniels) [inline src/auth/providers.js]: [q] Maybe this should be called customAuth or is that too repetitive?

File: test/test.js [hunk 4]
```diff
@@ -363,6 +365,41 @@ describe('anonymous auth', () => {
...
+
+    fetchMock.post(FUNCTION_CALL_URL, () => {
+      return JSON.stringify({x: ++count});
+    });
+  });
+
+  it('can authenticate with custom auth method', () => {
+    expect.assertions(3);
+    let testClient = new StitchClient('testapp');
+    return testClient.authenticate('custom', 'jwt')
```
Comment (edaniels) [inline test/test.js]: We should at least verify that the mocked endpoint receives `jwt`

File: src/auth/providers.js (hunk 3 above)
Comment (jsflax) [reply src/auth/providers.js]: Seems redundant and goes against the existing paradigm of not using `Auth` in the name

File: src/auth/providers.js (hunk 1 above)
Comment (jsflax) [reply src/auth/providers.js]: Ooops

Comment (mburdette) [reply src/auth/providers.js]: [nit] i think snake casing `userpass` seems weird

**Comment by edaniels** [comment]
bump @jsflax

**Comment by jsflax** [comment]
@edaniels PTAL

**Pull request merged**